src/shared/mount-util.c

```diff
@@ -475,67 +475,86 @@ int bind_remount_one_with_mountinfo(
+        if (in_initrd()) {
+                if (mount(path, "/", NULL, MS_MOVE, NULL) < 0)
+                        return log_debug_errno(errno, "Failed to move new rootfs '%s': %m", path);
 
+                if (chroot(".") < 0)
+                        return log_debug_errno(errno, "Failed to chroot to new rootfs '%s': %m", path);
 
-        /* At this point the new root is tucked under the old root. If we want
-         * to unmount it we cannot be fchdir()ed into it. So escape back to the
-         * old root. */
-        if (fchdir(fd_oldroot) < 0)
-                return log_debug_errno(errno, "Failed to change back to old rootfs: %m");
+                if (chdir("/"))
+                        return log_debug_errno(errno, "Failed to chdir to new rootfs '%s': %m", path);
+        } else {
+                _cleanup_close_ int fd_oldroot = -EBADF;
 
-        /* Note, usually we should set mount propagation up here but we'll
-         * assume that the caller has already done that. */
+                fd_oldroot = open("/", O_PATH | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+                if (fd_oldroot < 0)
+                        return log_debug_errno(errno, "Failed to open old rootfs");
 
-        /* Get rid of the old root and reveal our brand new root. */
-        if (umount2(".", MNT_DETACH) < 0)
-                return log_debug_errno(errno, "Failed to unmount old rootfs: %m");
+                /* Let the kernel tuck the new root under the old one. */
+                if (pivot_root(".", ".") < 0)
+                        return log_debug_errno(errno, "Failed to pivot root to new rootfs '%s': %m", path);
 
-        if (fchdir(fd_newroot) < 0)
-                return log_debug_errno(errno, "Failed to switch to new rootfs '%s': %m", path);
+                /* At this point the new root is tucked under the old root. If we want
+                 * to unmount it we cannot be fchdir()ed into it. So escape back to the
+                 * old root. */
+                if (fchdir(fd_oldroot) < 0)
+                        return log_debug_errno(errno, "Failed to change back to old rootfs: %m");
+
+                /* Note, usually we should set mount propagation up here but we'll
+                 * assume that the caller has already done that. */
+
+                /* Get rid of the old root and reveal our brand new root. */
+                if (umount2(".", MNT_DETACH) < 0)
+                        return log_debug_errno(errno, "Failed to unmount old rootfs: %m");
+
+                if (fchdir(fd_newroot) < 0)
+                        return log_debug_errno(errno, "Failed to switch to new rootfs '%s': %m", path);
+        }
+
+        /* Finally, let's establish the requested propagation type. */
+        if ((flags != 0) && mount(NULL, ".", NULL, flags | MS_REC, 0) < 0)
+                return log_debug_errno(errno, "Failed to turn new rootfs '%s' into %s mount: %m", prop_name, path);
```

Comment:
hmm, this swaps around the order of the MS_REC thing in case of MS_MOVE, is that intended?

Reply:
It is. This isn't an issue thought since the 3rd patch in the series ports the places that need changing which is only the mount tunnel.